pkg/controller/common/association/gc.go

```diff
@@ -0,0 +1,260 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package association
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/user"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	log = logf.Log.WithName("association")
+)
+
+const (
+	APIBasePath   = "/apis"
+	AllNamespaces = ""
+)
+
+type clientFactory func(baseConfig *rest.Config, gv schema.GroupVersion) (rest.Interface, error)
+
+// UsersGarbageCollector allows to remove unused Users. Users should be deleted as part of the association controllers
+// reconciliation loop. But without a Finalizer nothing prevent the associated resource to be removed while the
+// operator is not running.
+// This code is intended to be run during startup, before the controllers are started, to detect and delete such
+// orphaned resources.
+type UsersGarbageCollector struct {
+	// clientset is used to list potential orphaned secrets
+	clientset kubernetes.Interface
```

Comment:
I'm not sure I understand the need for being that generic here. We use generic clientset + mapper  + scheme, but the struct is specialized in deleting Users Secrets.

Can we not just use the usual controller-runtime client, that is already populated with all secrets + all apm/Kibana/ES resources because of watches, then iterate through all secrets having a particular label? It avoids additional API calls + memory consumption for something we already have in memory.

The registration process could be a function that lists existing ES/Kibana/APM resources and produces a "generic" map of all existing resources:

```
"kibanaassociation.k8s.elastic.co" -> NamespacedName -> struct{}{}   (map[string]map[types.NamespacedName]struct{})
"apmassociation.k8s.elastic.co" -> NamespacedName -> struct{}{}
```

The gc function would then:
- list all secrets of type users (from the cache)
- for each secret
    - for each entry in the registration map
        - check if the label exists in the secret, if it does, check if the NamespacedName (ie the Kibana resource referenced in that label) exists in the registration map. If not, cleanup the secret.

The only not-generic thing would be a function that populates the map above: we'd probably have one for ES, almost the same one for Kibana, almost the same one for APM. I think that would be fine.

I think it would remove the complexity of using the generic clientset and working with GVKinds.

WDYT?

Reply:
Sum up of what we have discussed offline:

clientset and REST client are used instead of the controller runtime client because the latter can't be used until it is initialized _(which is not the case here because GC is done before the manager is started)_ 

To use the controller-runtime client an option would be to start the garbage collector in a go routine and wait for it to be ready. It seems that `mgr.GetCache().WaitForCacheSync(...)` can be used to do that. But if the GC process is in its own go routine it raises the question of what to do in case of error ?

My feeling is that using the controller runtime client will drag its own complexity, not sure it worth it, but no strong opinion either...

Reply:
I'm fine with keeping the current version for now :)
If we ever has slow startup issues or large RAM usage we can still easily revisit with no breaking change.

Reply:
👍 I was wrongly assuming that the controller runtime client was tightly coupled with the cache, which is not the case. I have updated the PR so the controller runtime client is now used.